logdna/resource_archive.go

```diff
@@ -0,0 +1,278 @@
+package logdna
+
+import (
+	"context"
+	"encoding/json"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+)
+
+const archiveConfigID = "archive"
+
+type archiveConfig struct {
```

Comment:
Nit: we can use [embedded structs](https://golang.org/doc/effective_go#embedding) to improve the organization here. I drafted something up after finally figuring out how this code block is meant to map to a single nested object
```golang
"<field_name>": {
  Type: schema.TypeList,
  MaxItems: 1,
  Elem: &schema.Resource{
    Schema: map[string]*schema.Schema{
      "string_prop": {
        Type: schema.TypeString,
        Required: true,
      },
      // ... more property schema(s)
    },
},
```

Reply:
Thanks for looking into it! I was unable to get it to work for some reason but it looks like the structure is
```
resource "logdna_archive" "config" {
  integration = "gcs"
  gcs_config {
    bucket = "bucket"
    projectid = "id"
  }
}
```
I would argue there is more friction for the user than what can be done currently. 
```
resource "logdna_archive" "config" {
  integration = "gcs"
  bucket = "bucket"
  projectid = "id"
}
```
We are able to define the schema restrictions more accurately but I believe this is not absolutely crucial due to the API validation messages. 

I also think `gcs_config` should be enough information, the additional `integration` field is redundant. 

Happy to discuss more!
https://github.com/logdna/terraform-provider-logdna/pull/34#discussion_r714300080 Referencing my other comment here as well.

Reply:
```hcl
resource "logdna_archive" "config" {
  gcs_config {
    bucket = "bucket"
    projectid = "id"
  }
}
```

this actually seems pretty nice IMO, and would avoid needing the `integration` key for the user - we can just set it in the request when processing. we could also then use the schema to say "only one of `gcs_config`, `dos_config`, ..." etc.

I think nailing down the user-facing structure is most important, it will be harder to change that later versus abstracting the code itself. That said, what @david-ly put together is the direction I was alluding to before in my comment (embedded structs), and the schema changes there seem pretty reasonable